lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3541,19 +3548,14 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     //   }
     // }
     //
-    // Note: we relax the restriction if the type we're casting to is a
-    // non-final class because it's possible that we might have a subclass
-    // that conforms to the protocol.
-    if (fromExistential && !toExistential) {
-      if (auto NTD = toType->getAnyNominal()) {
-        if (!toType->is<ClassType>() || NTD->isFinal()) {
-          auto protocolDecl =
-              dyn_cast_or_null<ProtocolDecl>(fromType->getAnyNominal());
-          if (protocolDecl &&
-              !conformsToProtocol(toType, protocolDecl, dc)) {
-            return failed();
-          }
-        }
+    // But we check for conformance and possible dynamic conformances in which
+    // case we cannot fail because it is impossible to know then statically.
+    if (auto *protocolDecl =
+          dyn_cast_or_null<ProtocolDecl>(fromType->getAnyNominal())) {
+      // Checking for possible dynamic conformances because we cannot fail
+      // in those cases.
```

Comment:
These two comments appear to be saying mostly the same thing, could they be merged into one?

Reply:
In fact, it seems they could both be merged into the comment above as something along the lines of:

```
// If we have a cast from an existential type to a concrete type that we
// statically know doesn't conform to the protocol, mark the cast as always
// failing. For example:
```

What do you think?

Reply:
Looks great, I think the most detailed are already well commented on the `couldDynamicallyConformToProtocol` so this should be sufficient.